[flang][Frontend] Implement printing defined macros via -dM

kparzysz opened the pull request:
This should work the same way as in clang.

kparzysz committed:
[flang][Frontend] Implement printing defined macros via -dM

This should work the same way as in clang.

kparzysz commented:
We have a customer requesting information about macros defined by the compiler, and this feature would address the issue.

kparzysz commented:
cc: @bcornille

kparzysz commented:
The Windows build failure is due to an unrelated issue:
```
error LNK2019: unresolved external symbol _FortranAioSetForm referenced in function "private: virtual void __cdecl ExternalIOTests_TestWriteAfterEndfile_Test::TestBody(void)"
```

klausler commented on flang/include/flang/Parser/parsing.h:
```diff
@@ -59,6 +62,7 @@ class Parsing {
   const SourceFile *Prescan(const std::string &path, Options);
   void EmitPreprocessedSource(
       llvm::raw_ostream &, bool lineDirectives = true) const;
+  void EmitPreprocessorMacros(llvm::raw_ostream &) const;
```
Why are you dumping the macros from the parser?  Their definitions all come from the driver via Fortran::parser::Options::predefinitions -- except `_CUDA` which really should come from the drivers anyway.  Print them from the drivers instead.

klausler replied:
Ah, I see; you also want to dump macros that are defined in headers; never mind.

kparzysz replied:
We need to print macros that are defined in the source code.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
 
+void Definition::Print(
+    llvm::raw_ostream &out, llvm::StringRef macroName) const {
+  if (isDisabled_) {
```
Macros are only temporarily disabled during expansion; not relevant here.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
+  // The sequence of characters from which argument names will be created.
+  static llvm::StringRef charSeq{"ABCDEFGHIJKLMNOPQRSTUVWXYZ"};
+
+  auto couldCollide = [&](llvm::StringRef str) {
```
Always use braced initialization in the C++17 bits of f18, please.  It's safer.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
+  // For function-like macros we need to invent valid argument names (they
+  // are represented as ~A, ~B, ...). These invented names cannot collide
+  // with any other tokens in the macro definitions.
+  llvm::SmallSet<std::string, 10> usedNames;
```
You can use `std::set<>` here.  We avoid needless llvm dependencies in the parts of f18 that people are using to build Fortran tooling outside the llvm project.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
+      size_t idx{static_cast<size_t>(tok[1] - 'A')};
+      if (idx < argumentCount_) {
+        out << argNames[idx];
+        continue;
```
Use `else` clauses and avoid the `continue`; it's clearer to read.

kparzysz replied:
Refactored the ifs.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -713,6 +820,27 @@ void Preprocessor::Directive(const TokenSequence &dir, Prescanner &prescanner) {
+void Preprocessor::PrintMacros(llvm::raw_ostream &out) const {
+  // Sort the entries by macro name.
+  llvm::SmallVector<decltype(definitions_)::const_iterator> entries;
+  for (auto it{definitions_.begin()}, e{definitions_.end()}; it != e; ++it) {
```
A range-based `for` would be cleaner.

kparzysz replied:
I'm collecting iterators to sort them.

kparzysz replied:
Using std::set now.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -713,6 +820,27 @@ void Preprocessor::Directive(const TokenSequence &dir, Prescanner &prescanner) {
 
+void Preprocessor::PrintMacros(llvm::raw_ostream &out) const {
+  // Sort the entries by macro name.
+  llvm::SmallVector<decltype(definitions_)::const_iterator> entries;
```
Use an ordered `std::set<>` and you won't need to sort the result.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -713,6 +820,27 @@ void Preprocessor::Directive(const TokenSequence &dir, Prescanner &prescanner) {
+    return it1->first.ToString() < it2->first.ToString();
+  });
+
+  for (auto &&it : entries) {
```
`for (const auto &[name, def] : entries) {`

kparzysz replied:
Entries are iterators.

kparzysz replied:
Changed to use std:;set.

kparzysz commented on flang/include/flang/Parser/parsing.h:
```diff
@@ -83,6 +87,7 @@ class Parsing {
   const char *finalRestingPlace_{nullptr};
   std::optional<Program> parseTree_;
   ParsingLog log_;
+  std::unique_ptr<Preprocessor> preprocessor_;
```
Because `Preprocessor` is an incomplete class here, and the preprocessor header is not exported.

klausler replied:
Yes, but it could be, and then there's no question at use sites whether `preprocessor_->` can crash or not.

kparzysz replied:
It doesn't have a default constructor---it needs `AllSources`, which only becomes available in `Prescan`.  Do you want to add a default constructor, and a function `setAllSources`?

klausler replied:
`Preprocessor preprocessor_{allCookedSources_.allSources()};` should work.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
+    if (tok.empty()) {
+      continue;
+    }
+    // The generated names will only use characters from `charSeq`, so
```
You can avoid all of this complexity if you modify the preprocessor to retain the original actual argument names in the macro bodies.

kparzysz replied:
Yes, but I didn't want to make that change on my own.  Do you want me to go ahead with that?

kparzysz replied:
Do you want to store the argument list (with names) as well?  If not, we'd have to extract the names from the body, and sort them it based on the `~?~` character.

kparzysz replied:
I did something a bit different: I'm storing the argument names instead of making any changes to the tokenization.  The reason is that if there is a macro with an unused argument, we still need to give it some name.  Storing all of the actual names would completely eliminate the need to do that.

kparzysz committed:
Address review comments

kparzysz committed:
Move preprocessor.h and token-sequence.h to flang/include directory

preprocessor.h is moved to make definition of Preprocessor available
in parsing.h.
token-sequence is moved because preprocessor.h depends on it.

kparzysz committed:
Make Preprocessor not be dynamically allocated

kparzysz committed:
Store argument names, remove argument name generation

kparzysz committed:
clang-format

kparzysz committed:
clang-format

kparzysz committed:
clang-format

kparzysz commented:
The Windows build has the same failure as before (it's a known build breakage):
```
error LNK2019: unresolved external symbol _FortranAioSetForm referenced in function "private: virtual void __cdecl ExternalIOTests_TestWriteAfterEndfile_Test::TestBody(void)" (?TestBody@ExternalIOTests_TestWriteAfterEndfile_Test@@EEAAXXZ)
```

kparzysz committed:
Merge branch 'main' into user/kparzysz/flang-dm

klausler commented on flang/include/flang/Parser/preprocessor.h:
```diff
@@ -49,15 +51,21 @@ class Definition {
+  // For a given token, return the index of the argument to which the token
+  // corresponds, or `argumentCount` if the token does not correspond to any
+  // argument.
+  std::size_t getArgumentIndex(const CharBlock &token) const;
```
Capitalize function and member function names for consistency in the parser, please.

kparzysz replied:
Done

klausler commented on flang/include/flang/Parser/preprocessor.h:
```diff
@@ -49,15 +51,21 @@ class Definition {
 
   TokenSequence Apply(const std::vector<TokenSequence> &args, Prescanner &);
 
+  void Print(llvm::raw_ostream &out, llvm::StringRef macroName = "") const;
```
`const char *` would be a more portable type for `macroName` for those who use this code outside llvm.

kparzysz replied:
Done.

klausler commented on flang/lib/Parser/preprocessor.cpp:
```diff
@@ -46,6 +49,38 @@ bool Definition::set_isDisabled(bool disable) {
+
+  for (size_t i{0}, e{replacement_.SizeInTokens()}; i != e; ++i) {
+    std::string tok{replacement_.TokenAt(i).ToString()};
+    if (size_t idx = getArgumentIndex(tok); idx < argCount) {
```
`std::size_t idx {GetArgumentIndex(Tok)}; ...`

kparzysz replied:
Done.

kparzysz committed:
Address review comments

klausler commented:
> How is this code used outside of LLVM? Why do people want to use _without_ LLVM? Just curious.

There are people who want to construct tools to analyze or restructure Fortran programs outside of the context of compilation, and need to have access to a good parse tree or symbol table.  Unlike most other languages, Fortran is weirdly hard to lex & parse, and tool writers would rather use an existing parser than do a partial job with something custom.  So as a general rule, I avoid using llvm utilities in parsing and semantics (esp. parsing) when the facilities in `std::` are capable and portable, and try to encourage others to do the same.

clementval commented:
> How is this code used outside of LLVM? Why do people want to use _without_ LLVM? Just curious.

I can see language features in IDE like vscode being a good candidate.

kparzysz commented:
Is building LLVMSupport and maybe a few other components a barrier, or do they just not want to do that?

kparzysz committed:
Remove unnecessary #includes

kparzysz committed:
Remove unnecessary #includes

kkwli commented:
@kparzysz Is it supposed to also print the predefined macros? Or only the user-defined macros?
